src/puppetlabs/trapperkeeper/core.clj

```diff
@@ -143,19 +179,34 @@
       (when-let [error (:error shutdown-reason)]
         (throw error)))))
 
+(schema/defn bootstrap-with-app
+  "Bootstraps a Trapperkeeper application and runs it."
+  [app :- (schema/protocol app/TrapperkeeperApp)]
+  ;; This adds the TrapperkeeperApp instance to the tk-apps list, so that
+  ;; it can be referenced in a remote nREPL session, etc.
+  (swap! internal/tk-apps conj app)
+  (internal/register-sighup-handler)
+  (run-app app)
+  (swap! internal/tk-apps (partial remove #{app})))
+
 (schema/defn run
   "Bootstraps a trapperkeeper application and runs it.
   Blocks the calling thread until trapperkeeper is shut down.
   `cli-data` is expected to be a map constructed by parsing the CLI args.
   (see `parse-cli-args`)"
   [cli-data :- common/CLIData]
-  (let [app (boot-with-cli-data cli-data)]
-    ;; This adds the TrapperkeeperApp instance to the tk-apps list, so that
-    ;; it can be referenced in a remote nREPL session, etc.
-    (swap! internal/tk-apps conj app)
-    (internal/register-sighup-handler)
-    (run-app app)
-    (swap! internal/tk-apps (partial remove #{app}))))
+  (bootstrap-with-app (boot-with-cli-data cli-data)))
+
+(schema/defn run-with-config
```

Comment:
Contemplating whether it might be plausible to just add another arity to the existing `run` function instead, i.e.:

```
(schema/defn run
  ;; Update docstring to note that cli-data and config should "agree".
  ([cli-data :- common/CLIData]
   (run cli-data (config/parse-config-data cli-data)))
  ([cli-data :- common/CLIData
    config :- (schema/pred map?)]
   (let [app (boot-with-config cli-data config)]
     ...)))
 ```